src/ophyd_async/core/_device.py

```diff
@@ -60,23 +75,31 @@ def set_name(self, name: str):
         name:
             New name to set
         """
-
-        # Ensure self.log is recreated after a name change
-        if hasattr(self, "log"):
-            del self.log
-
         self._name = name
-        for attr_name, child in self.children():
-            child_name = f"{name}-{attr_name.rstrip('_')}" if name else ""
-            child.set_name(child_name)
-            child.parent = self
+        # Ensure self.log is recreated after a name change
+        self.log = LoggerAdapter(
+            getLogger("ophyd_async.devices"), {"ophyd_async_device_name": self.name}
+        )
+        for child_name, child in self.children():
+            _setup_child(self, child_name, child)
+
+    def __setattr__(self, name: str, value: Any) -> None:
+        if name == "parent":
+            if self.parent not in (value, None):
+                raise TypeError(
+                    f"Cannot set the parent of {self} to be {value}: "
+                    f"it is already a child of {self.parent}"
+                )
+        elif isinstance(value, Device):
+            _setup_child(self, name, value)
```

Comment:
This now breaks the use of the name parameter in signal initialiser because lines like
```
            self.test_signal = soft_signal_rw(int, 0, test_signal_name)
```
result in the desired signal name getting applied to the signal during construction, but then it is clobbered as soon as the assignment happens.

Reply:
Good to know, will try and work out a way to restore the old behaviour and add a test for it.

Reply:
Added in [be6d103](https://github.com/bluesky/ophyd-async/pull/594/commits/be6d10326f921d7b80fe7c51fcd6f4b7ea93dab7)